controllers/extension-certificate-service/pkg/controller/certservice/actuator.go

```diff
@@ -0,0 +1,335 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package certservice
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"path/filepath"
+
+	"github.com/gardener/gardener-extensions/controllers/extension-certificate-service/pkg/controller/certservice/internal"
+
+	corev1 "k8s.io/api/core/v1"
+
+	"k8s.io/apiserver/pkg/authentication/user"
+
+	"github.com/gardener/gardener/pkg/utils/secrets"
+
+	"github.com/gardener/gardener-extensions/controllers/extension-certificate-service/pkg/imagevector"
+
+	"github.com/gardener/gardener/pkg/chartrenderer"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/yaml"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+
+	"sigs.k8s.io/controller-runtime/pkg/runtime/log"
+
+	"github.com/gardener/gardener-extensions/pkg/controller"
+
+	"github.com/gardener/gardener-extensions/controllers/extension-certificate-service/pkg/utils"
+	"github.com/gardener/gardener-extensions/pkg/util"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+
+	"github.com/go-logr/logr"
+
+	configv1alpha1 "github.com/gardener/gardener-extensions/controllers/extension-certificate-service/pkg/apis/config/v1alpha1"
+	"github.com/gardener/gardener-extensions/pkg/controller/extension"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	gardenv1beta1 "github.com/gardener/gardener/pkg/apis/garden/v1beta1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ActuatorName is the name of the Certificate Service actuator.
+const ActuatorName = "certificate-service-actuator"
+
+// NewActuator returns an actuator responsible for Extension resources.
+func NewActuator(config configv1alpha1.CertificateServiceConfiguration) extension.Actuator {
+	return &actuator{
+		logger:            log.Log.WithName(ActuatorName),
+		certServiceConfig: config,
+	}
+}
+
+type actuator struct {
+	client client.Client
+	config *rest.Config
+
+	certServiceConfig configv1alpha1.CertificateServiceConfiguration
+
+	logger logr.Logger
+}
+
+// CreateOrUpdate the Extension resource.
+func (a *actuator) CreateOrUpdate(ctx context.Context, ex *extensionsv1alpha1.Extension) error {
+	namespace := ex.GetNamespace()
+	cluster, err := controller.GetCluster(ctx, a.client, namespace)
+	if err != nil {
+		return err
+	}
+
+	if err := a.createCertBroker(ctx, cluster.Shoot, namespace); err != nil {
```

Comment:
Shouldn't this be the last step (first creating all requirements (kubeconfig, rbac)), and when everything worked we finally deploy the component?

Reply:
You are absolutely right! Thank you